sdk/eventhub/azure-eventhubs/azure/eventhub/aio/eventprocessor/_ownership_manager.py

```diff
@@ -23,74 +21,79 @@ class OwnershipManager(object):
         """
         self.cached_parition_ids = await self.eventhub_client.get_partition_ids()
 
-    async def _balance_ownership(self, all_partition_ids):
+    async def _balance_ownership(self, ownership_list, all_partition_ids):
         """Balances and claims ownership of partitions for this EventProcessor.
-        The balancing algorithm is:
-        1. Find partitions with inactive ownership and partitions that haven never been claimed before
-        2. Find the number of active owners, including this EventProcessor, for all partitions.
-        3. Calculate the average count of partitions that an owner should own.
-        (number of partitions // number of active owners)
-        4. Calculate the largest allowed count of partitions that an owner can own.
-        math.ceil(number of partitions / number of active owners).
-        This should be equal or 1 greater than the average count
-        5. Adjust the number of partitions owned by this EventProcessor (owner)
-            a. if this EventProcessor owns more than largest allowed count, abandon one partition
-            b. if this EventProcessor owns less than average count, add one from the inactive or unclaimed partitions,
-            or steal one from another owner that has the largest number of ownership among all owners (EventProcessors)
-            c. Otherwise, no change to the ownership
-
-        The balancing algorithm adjust one partition at a time to gradually build the balanced ownership.
-        Ownership must be renewed to keep it active. So the returned result includes both existing ownership and
-        the newly adjusted ownership.
-        This method balances but doesn't claim ownership. The caller of this method tries to claim the result ownership
-        list. But it may not successfully claim all of them because of concurrency. Other EventProcessors may happen to
-        claim a partition at that time. Since balancing and claiming are run in infinite repeatedly,
-        it achieves balancing among all EventProcessors after some time of running.
-
-        :return: List[Dict[str, Any]], A list of ownership.
         """
-        ownership_list = await self.partition_manager.list_ownership(
-            self.eventhub_name, self.consumer_group_name
-        )
+
         now = time.time()
         ownership_dict = {x["partition_id"]: x for x in ownership_list}  # put the list to dict for fast lookup
         not_owned_partition_ids = [pid for pid in all_partition_ids if pid not in ownership_dict]
-        timed_out_partition_ids = [ownership["partition_id"] for ownership in ownership_list
-                                   if ownership["last_modified_time"] + self.ownership_timeout < now]
+        timed_out_partitions = [x for x in ownership_list
+                                if x["last_modified_time"] + self.ownership_timeout < now]
+        if self._initializing:  # greedily claim all available partitions when an EventProcessor is started.
```

Comment:
How thread safe is this? If customer wants to use the same client in multiple thread, could we have a problem of multiple initializtion?